timestream.lua

```diff
@@ -0,0 +1,393 @@
+-- speeds up the calendar, units, or both
+
+--[====[
+
+timestream
+==========
+This only affects fortress mode.
+
+Usage:
+:timestream <scalar> <fps> <simulate units y/n>
+Examples:
+:timestream     2:          Calendar runs at x2 normal speed
+:timestream    -1 100:      Calendar runs at dynamic speed to simulate 100 FPS
+:timestream    -1 100 1:    Calendar & units are simulated at 100 FPS
+
+Original timestream.lua: https://gist.github.com/IndigoFenix/cf358b8c994caa0f93d5
+]====]
+
+
+local MINIMAL_FPS           = 10    -- This ensures you won't get crazy values on pausing/saving, or other artefacts on extremely low FPS.
+local DEFAULT_MAX_FPS       = 100
+
+
+--- DO NOT CHANGE BELOW UNLESS YOU KNOW WHAT YOU'RE DOING ---
+
+args={...}
+local rate=tonumber(args[1])
+local desired_fps = tonumber(args[2])
+local simulating_units = tonumber(args[3])  -- Setting this to 2 instead of 1 will use debug_turbospeed instead of adjusting the timers of all creatures. I don't quite like it because it makes everyone move like insects.
+local debug_mode = tonumber(args[4])
```

Comment:
If these arguments are not all required, I might actually recommend using `utils.processArgs` instead. It would allow script invocations like this:
```
timestream -fps 200
timestream -rate 2
timestream -rate 3 -units
```
Arguments can be passed in without values, like `-units` or `-debug` - you would still need to convert from strings to the desired integer/boolean types manually, but I think it would be easier to handle this way, and easier to use. Searching for "processArgs" in this repo should turn up a lot of examples, but let me know if you have any questions about it.

Reply:
> Just saw your change and managed to test a bit on my end - sorry for the delay there.
> 
> I'm testing on a fresh map and occasionally seeing an issue where tree-chopping designations created after running `timestream 5` or `timestream 5 -1 1` are not done by any dwarves at all. Rerunning timestream with a different speedup (like 4 or 6) causes the jobs to be done immediately. My first guess was that the `process_dig` global wasn't being set or cleared by DF properly, but it seems to be handled as expected. Do you have any ideas as to what could be causing this? I'm not really sure where to start looking.
> 
> I was partly reminded by your most recent commit - are you still testing this? I'm fine with waiting if you're still performing some tests, but can also merge this whenever it's ready.

This might be related to the original `timestream` bug where units are strictly preoccupied with leisure activities and therefore don't do anything else. This is why the original script didn't work after the tavern system was introduced. I thought I fixed this bug because I haven't encountered it since, but frankly I have only tried messing around with the regular calendar speeding up only a handful of times, and strictly tested the FPS simulation. I haven't tested what could happen at extremely high values, either.

As for the main event, aka the FPS simulation; I've been running a fort for 70 years now (longest I've ever had, would've quit due to the FPS death otherwise) and found no significant bugs other than the ones I pointed out in the main post, which are not gamebreaking by any means, so to me I think it's good to go. I haven't tested the plain portion of the script, though.

Reply:
Yeah, I wasn't sure how extreme my test was. I'll see if I have a low-FPS fort sitting around, but that sounds good to me.

Any thoughts on the argument handling comment? The main issue I can think of with `processArgs` is that it doesn't handle negative numbers well, but you seem to just be using them to indicate "unspecified" here, which wouldn't be necessary with `processArgs`.

Reply:
I'll implement it when I get the time.

Reply:
Would you be okay with me changing argument handling after I merge? Asking just in case you've already started or have something else in mind.

Reply:
> Would you be okay with me changing argument handling after I merge? Asking just in case you've already started or have something else in mind.

(Un?)fortunately, I haven't had a lot of spare time recently so I haven't begun to change it. You have my gratitude.